MaxText/tests/gradient_accumulation_test.py

```diff
@@ -0,0 +1,72 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+     https://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+"""
+
+# pylint: disable=missing-module-docstring, missing-function-docstring
+import numpy as np
+import json
+import unittest
+import pytest
+import string
+import random
+from train import main as train_main
+
+def generate_random_string(length=10):
+    characters = string.ascii_letters  # Include letters, digits, and punctuation
+    return ''.join(random.choice(characters) for _ in range(length))
+
+class GradientAccumulationTest(unittest.TestCase):  
+
+  
+  @pytest.mark.tpu
+  def test_grad_accumulate_same_loss(self):
+    random_suffix = generate_random_string()
+    run_accumulate_metrics_file = f"/tmp/runner_grad_accumulate_{random_suffix}.txt"
+    print(f"{run_accumulate_metrics_file=}")
+    run_regular_metrics_file = f"/tmp/runner_regular_{random_suffix}.txt"
+    print(f"{run_regular_metrics_file=}")
+    shared_maxtext_args = [
+       None,
+        "configs/base.yml",
+          r"base_output_directory=gs://runner-maxtext-logs",
+          r"dataset_path=gs://maxtext-dataset",
+          "enable_checkpointing=False",
+          "base_emb_dim=256",
+          "base_num_decoder_layers=4",
+          "tokenizer_path=../assets/tokenizer.llama2",
+          "steps=50"
+    ]
+    # Run with gradient accumulation with accumulate_steps=10, per_device_batch=1 --> simulating per_device_batch=10
+    train_main(shared_maxtext_args + [
+          "run_name=runner_grad_accumulate",
+          f"metrics_file={run_accumulate_metrics_file}",
+          "per_device_batch_size=1",
+          "gradient_accumulation_steps=10",       
+    ])
+
+    #Run without gradient accumulation with per_device_batch=10
+    train_main(shared_maxtext_args + [
+          "run_name=runner_grad_accumulate_regular",
+          f"metrics_file={run_regular_metrics_file}",
+          "per_device_batch_size=10",
+          "gradient_accumulation_steps=1",       
+    ])
+
+    with open(run_accumulate_metrics_file, 'r', encoding='utf8') as accum_run,\
+          open(run_regular_metrics_file, 'r', encoding='utf8') as regular_run:
+        accum_run_loss = json.loads(accum_run.readlines()[-1])["learning/loss"]
+        regular_run_loss = json.loads(regular_run.readlines()[-1])["learning/loss"]
+        print(f"[Gradient Accumulation Test] Loss with gradient accumulation: {accum_run_loss}", flush=True)
+        print(f"[Gradient Accumulation Test] Loss without gradient accumulation: {regular_run_loss}", flush=True)
+        np.testing.assert_allclose(accum_run_loss, regular_run_loss, rtol=0.05)
```

Comment:
Wondering if the flops from `calculate_tflops_training_per_device` can be captured somewhere?

Reply:
Added an assert for this!